File: src/server/backend/runtime/checkers/manager.py
"""Server checker manager: phased checker execution."""
from __future__ import annotations

import importlib
import inspect
import json
from pathlib import Path
from typing import Any

from shared.schemas.context import RuntimeContext
from shared.schemas.events import EventType, RuntimeEvent
from backend.runtime.checkers.base import BaseChecker, CheckResult
from backend.runtime.checkers.registry import get_checker_class

PHASE_ORDER = ("llm_before", "llm_after", "tool_before", "tool_after", "global")

_EVENT_PHASE = {
    EventType.LLM_INPUT: "llm_before",
    EventType.LLM_OUTPUT: "llm_after",
    EventType.TOOL_INVOKE: "tool_before",
    EventType.TOOL_RESULT: "tool_after",
}

def default_checkers() -> list[BaseChecker]:
    return []


def default_checker_config() -> dict[str, dict[str, list[Any]]]:
    return {}


def load_checker_config(source: str | Path | dict[str, Any] | None) -> dict[str, list[Any]]:
    if source is None:
        return {}
    elif isinstance(source, (str, Path)):
        path = Path(source)
        with path.open("r", encoding="utf-8") as fh:
            data = json.load(fh)
    else:
        data = dict(source)

    phases = data.get("phases")
    if not isinstance(phases, dict):
        raise ValueError("checker config must contain a 'phases' object")
    config: dict[str, list[Any]] = {}
    for phase in PHASE_ORDER:
        if phase in phases:
            config[phase] = _checker_specs_for_scope(phases.get(phase), "remote")
    return config


def _checker_specs_for_scope(value: Any, scope: str) -> list[Any]:
    if not isinstance(value, dict):
        raise ValueError("checker phase config must be an object with 'local' and 'remote'")
    if "local" not in value or "remote" not in value:
        raise ValueError("checker phase config must include both 'local' and 'remote'")
    specs = value.get(scope)
    if specs is None:
        return []
    if not isinstance(specs, list):
        raise ValueError(f"checker phase '{scope}' config must be a list")
    return list(specs)


def build_checkers_by_phase(config: dict[str, list[Any]]) -> dict[str, list[BaseChecker]]:
    return {
        phase: [_instantiate_checker(spec) for spec in specs]
        for phase, specs in config.items()
    }


class CheckerManager:
    """Runs configured checkers for the event phase and merges CheckResults."""

    def __init__(
        self,
        checkers: list[BaseChecker] | None = None,
        *,
        config: str | Path | dict[str, Any] | None = None,
    ) -> None:
        if checkers is not None:
            self.checkers_by_phase = {"global": list(checkers)}
        else:
            self.checkers_by_phase = build_checkers_by_phase(load_checker_config(config))
        self._refresh_flat_checkers()

    def update_config(self, config: str | Path | dict[str, Any] | None) -> None:
        """Replace checker configuration for subsequent server decisions."""
        self.checkers_by_phase = build_checkers_by_phase(load_checker_config(config))
        self._refresh_flat_checkers()

    def _refresh_flat_checkers(self) -> None:
        self.checkers = [
            checker
            for phase in PHASE_ORDER
            for checker in self.checkers_by_phase.get(phase, [])
        ]

    def add(self, checker: BaseChecker, phase: str | None = None) -> None:
        target = phase or _infer_phase(checker)
        self.checkers_by_phase.setdefault(target, []).append(checker)
        self.checkers.append(checker)

    def run(
        self,
        event: RuntimeEvent,
        context: RuntimeContext,
        *,
        trajectory_window: list[RuntimeEvent] | None = None,
        stop_on_first_decision: bool = False,
    ) -> CheckResult:
        merged_signals: list[str] = []
        candidate = None
        is_final = False
        meta: dict = {}
        phase = _EVENT_PHASE.get(event.event_type, "global")
        phase_checkers = list(self.checkers_by_phase.get(phase, []))
        phase_checkers.extend(self.checkers_by_phase.get("global", []))
        for checker in phase_checkers:
            if not checker.applies(event):
                continue
            try:
                res = _call_checker(checker, event, context, trajectory_window)
            except Exception as exc:
                meta[f"{checker.name}_error"] = str(exc)
                continue
            for signal in res.risk_signals:
                if signal not in merged_signals:
                    merged_signals.append(signal)
            if res.metadata:
                meta.update(res.metadata)
            if res.decision_candidate and (candidate is None or res.is_final):
                candidate = res.decision_candidate
                is_final = is_final or res.is_final
                if stop_on_first_decision:
                    break

        for signal in merged_signals:
            event.add_signal(signal)
        return CheckResult(
            decision_candidate=candidate,
            risk_signals=merged_signals,
            is_final=is_final,
            metadata=meta,
        )


def _instantiate_checker(spec: Any) -> BaseChecker:
    if isinstance(spec, BaseChecker):
        return spec
    if isinstance(spec, type) and issubclass(spec, BaseChecker):
        return spec()
    if isinstance(spec, str):
        cls = get_checker_class(spec) or _load_checker_class(spec)
        return cls()
    if isinstance(spec, dict):
        target = spec.get("class") or spec.get("checker") or spec.get("name")
        kwargs = dict(spec.get("kwargs") or {})
        if isinstance(target, str):
            cls = get_checker_class(target) or _load_checker_class(target)
        elif isinstance(target, type) and issubclass(target, BaseChecker):
            cls = target
        else:
            raise ValueError(f"invalid checker config entry: {spec!r}")
        return cls(**kwargs)
    raise ValueError(f"invalid checker config entry: {spec!r}")


def _call_checker(
    checker: BaseChecker,
    event: RuntimeEvent,
    context: RuntimeContext,
    trajectory_window: list[RuntimeEvent] | None,
) -> CheckResult:
    """Call new trace-aware checkers while tolerating old two-arg checkers."""
    params = inspect.signature(checker.check).parameters
    accepts_trace = any(p.kind == inspect.Parameter.VAR_POSITIONAL for p in params.values())
    accepts_trace = accepts_trace or len(params) >= 3
    if accepts_trace:
        return checker.check(event, context, trajectory_window)
    return checker.check(event, context)  # type: ignore[call-arg]


def _load_checker_class(path: str) -> type[BaseChecker]:
    module_name, _, class_name = path.rpartition(".")
    if not module_name or not class_name:
        raise ValueError(f"checker must be a builtin name or import path: {path}")
    module = importlib.import_module(module_name)
    cls = getattr(module, class_name)
    if not isinstance(cls, type) or not issubclass(cls, BaseChecker):
        raise TypeError(f"checker class must subclass BaseChecker: {path}")
    return cls


def _infer_phase(checker: BaseChecker) -> str:
    for event_type in checker.event_types:
        phase = _EVENT_PHASE.get(event_type)
        if phase:
            return phase
    return "global"

File: src/client/js/agentguard/checkers/manager.js
"use strict";

const fs = require("fs");
const { CheckResult, BaseChecker } = require("./base");
const { getCheckerClass, discoverCheckers } = require("./registry");
const { LLMInputChecker } = require("./llm_before/llm_input");
const { LLMOutputChecker } = require("./llm_after/llm_output");
const { ToolInvokeChecker } = require("./tool_before/tool_invoke");
const { ToolResultChecker } = require("./tool_after/tool_result");

const PHASE_ORDER = ["llm_before", "llm_after", "tool_before", "tool_after", "global"];
const EVENT_PHASE = {
  llm_input: "llm_before",
  llm_output: "llm_after",
  tool_invoke: "tool_before",
  tool_result: "tool_after",
};
const BUILTIN_CHECKERS = {
  llm_input: LLMInputChecker,
  llm_output: LLMOutputChecker,
  tool_invoke: ToolInvokeChecker,
  tool_result: ToolResultChecker,
};

function defaultCheckers() {
  return [];
}

function loadCheckerConfig(source = null) {
  if (source == null) {
    return null;
  }
  let data;
  if (typeof source === "string") {
    data = JSON.parse(fs.readFileSync(source, "utf-8"));
  } else {
    data = { ...source };
  }
  const phases = data.phases;
  if (!phases || typeof phases !== "object" || Array.isArray(phases)) {
    throw new Error("checker config must contain a 'phases' object");
  }
  const config = {};
  for (const phase of PHASE_ORDER) {
    if (phase in phases) {
      config[phase] = checkerSpecsForScope(phases[phase], "local");
    }
  }
  return config;
}

function checkerSpecsForScope(value, scope) {
  if (!value || typeof value !== "object" || Array.isArray(value)) {
    throw new Error("checker phase config must be an object with 'local' and 'remote'");
  }
  if (!("local" in value) || !("remote" in value)) {
    throw new Error("checker phase config must include both 'local' and 'remote'");
  }
  const specs = value[scope];
  if (specs == null) {
    return [];
  }
  if (!Array.isArray(specs)) {
    throw new Error(`checker phase '${scope}' config must be a list`);
  }
  return [...specs];
}

function buildCheckersByPhase(config = null) {
  if (!config) {
    return {};
  }
  const result = {};
  for (const [phase, specs] of Object.entries(config)) {
    result[phase] = specs.map(instantiateChecker);
  }
  return result;
}

function instantiateChecker(spec) {
  if (spec instanceof BaseChecker) {
    return spec;
  }
  if (typeof spec === "function") {
    return buildChecker(spec);
  }
  if (typeof spec === "string") {
    discoverCheckers();
    const CheckerClass = BUILTIN_CHECKERS[spec] || getCheckerClass(spec);
    if (!CheckerClass) {
      throw new Error(`invalid checker config entry: ${String(spec)}`);
    }
    return buildChecker(CheckerClass);
  }
  if (spec && typeof spec === "object") {
    const target = spec.class || spec.checker || spec.name;
    const kwargs = checkerKwargs(spec);
    const env = checkerEnv(spec);
    const CheckerClass = typeof target === "function" ? target : BUILTIN_CHECKERS[target] || getCheckerClass(target);
    if (!CheckerClass) {
      throw new Error(`invalid checker config entry: ${JSON.stringify(spec)}`);
    }
    return buildChecker(CheckerClass, { kwargs, env });
  }
  throw new Error(`invalid checker config entry: ${String(spec)}`);
}

function checkerKwargs(spec) {
  const reserved = new Set(["class", "checker", "name", "kwargs", "env"]);
  const kwargs = Object.fromEntries(Object.entries(spec).filter(([key]) => !reserved.has(key)));
  if (spec.kwargs != null && (typeof spec.kwargs !== "object" || Array.isArray(spec.kwargs))) {
    throw new Error(`checker kwargs config must be an object: ${JSON.stringify(spec)}`);
  }
  return { ...kwargs, ...(spec.kwargs || {}) };
}

function checkerEnv(spec) {
  if (spec.env != null && (typeof spec.env !== "object" || Array.isArray(spec.env))) {
    throw new Error(`checker env config must be an object: ${JSON.stringify(spec)}`);
  }
  return { ...(spec.env || {}) };
}

function buildChecker(CheckerClass, { kwargs = null, env = null } = {}) {
  const checkerKwargs = { ...(kwargs || {}) };
  const checkerEnv = { ...(env || {}) };
  try {
    return new CheckerClass({ env: checkerEnv, ...checkerKwargs });
  } catch (_) {
    const checker = new CheckerClass();
    if (typeof checker.bind_config === "function") {
      checker.bind_config({ env: checkerEnv, ...checkerKwargs });
    }
    return checker;
  }
}

class CheckerManager {
  constructor({ checkers = null, config = null } = {}) {
    this.checkers_by_phase = checkers ? { global: [...checkers] } : buildCheckersByPhase(loadCheckerConfig(config));
    this.refresh();
  }

  update_config(config = null) {
    this.checkers_by_phase = buildCheckersByPhase(loadCheckerConfig(config));
    this.refresh();
  }

  updateConfig(config = null) {
    this.update_config(config);
  }

  add(checker, phase = null) {
    const target = phase || inferPhase(checker);
    this.checkers_by_phase[target] = this.checkers_by_phase[target] || [];
    this.checkers_by_phase[target].push(checker);
    this.checkers.push(checker);
  }

  refresh() {
    this.checkers = PHASE_ORDER.flatMap((phase) => this.checkers_by_phase[phase] || []);
  }

  run(event, context) {
    const phase = EVENT_PHASE[event.event_type] || "global";
    const phaseCheckers = [...(this.checkers_by_phase[phase] || []), ...(this.checkers_by_phase.global || [])];
    const mergedSignals = [];
    let candidate = null;
    let isFinal = false;
    const metadata = {};
    for (const checker of phaseCheckers) {
      if (!checker.applies(event)) {
        continue;
      }
      try {
        const result = checker.check(event, context);
        for (const signal of result.risk_signals) {
          if (!mergedSignals.includes(signal)) {
            mergedSignals.push(signal);
          }
        }
        Object.assign(metadata, result.metadata || {});
        if (result.decision_candidate && (candidate === null || result.is_final)) {
          candidate = result.decision_candidate;
          isFinal = isFinal || result.is_final;
        }
      } catch (error) {
        metadata[`${checker.name}_error`] = String(error.message || error);
      }
    }
    for (const signal of mergedSignals) {
      event.addSignal(signal);
    }
    return new CheckResult({
      decision_candidate: candidate,
      risk_signals: mergedSignals,
      is_final: isFinal,
      metadata,
    });
  }
}

function inferPhase(checker) {
  for (const eventType of checker.event_types || []) {
    const phase = EVENT_PHASE[eventType];
    if (phase) {
      return phase;
    }
  }
  return "global";
}

module.exports = {
  PHASE_ORDER,
  CheckerManager,
  defaultCheckers,
  loadCheckerConfig,
  load_checker_config: loadCheckerConfig,
};
